### PR TITLE
[Merged by Bors] - Avoid pulling poet proof multiple times in 1:N setups

### PR DESCRIPTION
**Pull request by poszu**
## Motivation

All identities within the node register in the same poets so there's no need to fetch the proof many times.

## Description

- extracted a higher-level interface `PoetClient` that:
  - abstracts away PoW
  - avoids fetching a proof many times by serializing with a mutex and caching
  - will be responsible for poet certification in the future

## Test Plan

Added test for caching functionality

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

**Comment by poszu** [comment]
bors try

**Comment by poszu** [comment]
bors try

File: activation/poetdb.go [hunk 1]
```diff
@@ -156,6 +156,16 @@ func (db *PoetDb) GetProof(proofRef types.PoetProofRef) (*types.PoetProof, *type
 	return &proofMessage.PoetProof, &proofMessage.Statement, nil
 }
 
+func (db *PoetDb) GetProofForRound(poetID []byte, roundID string) (*types.PoetProof, error) {
```
Comment (fasmat) [inline activation/poetdb.go]: NIT:

```suggestion
func (db *PoetDb) ProofForRound(poetID []byte, roundID string) (*types.PoetProof, error) {
```

File: activation/poet.go [hunk 2]
```diff
@@ -272,3 +268,112 @@ func (c *HTTPPoetClient) req(ctx context.Context, method, path string, reqBody,
...
+	}
+
+	getProofsCtx, cancel := withConditionalTimeout(ctx, c.requestTimeout)
+	defer cancel()
+	proof, members, err := c.client.Proof(ctx, roundID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting proof: %w", err)
+	}
+
+	if err := c.db.ValidateAndStore(getProofsCtx, proof); err != nil && !errors.Is(err, ErrObjectExists) {
```
Comment (fasmat) [inline activation/poet.go]: The timeout should only be for the request to poet, not for storing the result (compare with previous code).

```suggestion
	if err := c.db.ValidateAndStore(ctx, proof); err != nil && !errors.Is(err, ErrObjectExists) {
```

File: activation/poet_client_test.go [hunk 3]
```diff
@@ -138,26 +142,48 @@ func Test_HTTPPoetClient_Proof(t *testing.T) {
 	require.NoError(t, err)
 }
 
-func Test_HTTPPoetClient_PoetServiceID(t *testing.T) {
+func TestPoetClient_CachesProof(t *testing.T) {
+	t.Parallel()
+
+	proofsCalled := atomic.Uint64{}
```
Comment (fasmat) [inline activation/poet_client_test.go]:
```suggestion
	var proofsCalled atomic.Uint64
```

**Review by fasmat — APPROVED** [review]

**Comment by poszu** [comment]
bors merge

**Comment by poszu** [comment]
Bors merge

**Pull request closed without merging**